common/src/tor.rs

```diff
@@ -0,0 +1,199 @@
+// following disclaimer in the documentation and/or other materials provided with the distribution.
+//
+// 3. Neither the name of the copyright holder nor the names of its contributors may be used to endorse or promote
+// products derived from this software without specific prior written permission.
+//
+// THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES,
+// INCLUDING, BUT NOT LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE
+// DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,
+// SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR
+// SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY,
+// WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE
+// USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+
+use crate::{exit_codes::ExitCodes, CommsTransport, TorControlAuthentication};
+use libtor::{LogDestination, LogLevel, TorFlag};
+use log::*;
+use multiaddr::Multiaddr;
+use pgp::{crypto::HashAlgorithm, types::StringToKey};
+use rand::{distributions::Alphanumeric, thread_rng, Rng};
+use std::{io, net::TcpListener, ops::Range};
+use tari_shutdown::ShutdownSignal;
+use tempfile::{tempdir, NamedTempFile, TempDir, TempPath};
+
+const LOG_TARGET: &str = "common::tor";
+
+pub struct Tor {
+    control_port: u16,
+    data_dir: String,
+    log_destination: String,
+    log_level: LogLevel,
+    passphrase: Option<String>,
+    socks_port: u16,
+    temp_dir: Option<TempDir>,
+    temp_file: Option<TempPath>,
+}
+
+impl Default for Tor {
+    fn default() -> Tor {
+        Tor {
+            control_port: 19_051,
+            data_dir: "/tmp/tor-data".into(),
+            log_destination: "/tmp/tor.log".into(),
+            log_level: LogLevel::Err,
+            passphrase: None,
+            socks_port: 19_050,
+            temp_dir: None,
+            temp_file: None,
+        }
+    }
+}
+
+impl Tor {
+    /// Returns a new Tor instance with random options.
+    /// The data directory, passphrase, and log destination are temporary and randomized.
+    /// Two available adjacent TCP ports will be selected in the given
+    /// port range. These are scanned sequentially from start to end.
+    /// These ports are only for the control and socks ports, the onion address and port info are still loaded from the
+    /// node identity file.
+    pub fn randomize(port_range: Range<u16>) -> Result<Tor, ExitCodes> {
+        let mut instance = Tor::default();
+
+        // check for unused ports to assign
+        let (socks_port, control_port) = get_available_ports(port_range)?;
+        instance.socks_port = socks_port;
```

Comment:
If you set both of these to `0` the OS will assign a port.

Reply:
In that case does the bind return the port number?

Reply:
Okay I see I can use local_addr to get it from the listener. Will do